aws/resource_aws_acm_certificate_test.go

```diff
@@ -0,0 +1,311 @@
+package aws
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/acm"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+	"os"
+	"regexp"
+)
+
+func TestAccAwsAcmResource_certificateIssuingFlow(t *testing.T) {
+	var conf acm.DescribeCertificateOutput
+	var tags acm.ListTagsForCertificateOutput
+	var confAfterValidation acm.DescribeCertificateOutput
+	if os.Getenv("ACM_CERTIFICATE_ROOT_DOMAIN") == "" {
+		t.Skip("Environment variable ACM_CERTIFICATE_ROOT_DOMAIN is not set")
+	}
+
+	root_zone_domain := os.Getenv("ACM_CERTIFICATE_ROOT_DOMAIN")
+	domain := fmt.Sprintf("certtest.%s", root_zone_domain)
+	sanDomain := fmt.Sprintf("certtest2.%s", root_zone_domain)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAcmCertificateDestroy,
+		Steps: []resource.TestStep{
+			// Test that we can request a certificate
+			resource.TestStep{
+				Config: testAccAcmCertificateConfig(domain, sanDomain),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAcmCertificateExists("aws_acm_certificate.cert", &conf, &tags),
+					testAccCheckAcmCertificateAttributes("aws_acm_certificate.cert", &conf, domain, sanDomain, "PENDING_VALIDATION"),
```

Comment:
I found this pattern in some other tests, e.g. `testAccCheckAWSENIAttributes`. 
I guess it's a bit more paranoid in the sense that it checks the actual state in AWS without relying on the read operation. In theory, a no-op create and read would also result in the correct resource attributes in the state since they are initially populated from the resource declaration in the TF config, right? 

Is that worth the extra complexity? I'm not sure...
For now I added the checks you suggested, they definitely make the the expectations more explicit.

Reply:
Generally speaking we rely on the resource state checks (such as `resource.TestCheckResourceAttr`) as the acceptance testing is already paranoid about refreshing the resources from AWS. For many AWS services, having the _additional_ checks makes the acceptance testing more susceptible to eventual consistency issues. Some of the older provider testing code may have predated the state helpers or just was never converted over.

I would really consider getting rid of these unless there's some reason to keep them, which generally means an attribute should be added.

Reply:
Done, removed the custom checks. I don't think there's anything missing except the check that the certificate is in state `ISSUED` but the validation resource checks that.